tooling/templatize/pkg/aks/admin.go

```diff
@@ -0,0 +1,181 @@
+package aks
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	armauthorization "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v3"
+	"github.com/google/uuid"
+	auth "github.com/microsoft/kiota-authentication-azure-go"
+	msgraphsdk "github.com/microsoftgraph/msgraph-sdk-go"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	clusterAdminRoleID = "b1ff04bb-8a4e-4dc4-8eb5-8693973ce19b" // Azure Kubernetes Service RBAC Cluster Admin
+)
+
+type ClusterAdminAssignmentOptions struct {
+	Timeout        time.Duration
+	CheckFrequency time.Duration
+}
+
+func EnsureClusterAdmin(ctx context.Context, kubeconfigPath, subscriptionID, resourceGroupName, aksClusterName string, options *ClusterAdminAssignmentOptions) error {
+	if options == nil {
+		options = &ClusterAdminAssignmentOptions{
+			Timeout:        time.Duration(2 * time.Minute),
+			CheckFrequency: time.Duration(5 * time.Second),
+		}
+	}
+
+	// Get the current user's object ID
+	userObjectID, err := getCurrentUserObjectID(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get current user object ID: %w", err)
+	}
+
+	// Assign the Azure Kubernetes Service RBAC Cluster Admin role to the current user
+	err = assignClusterAdminRBACRole(ctx, subscriptionID, resourceGroupName, aksClusterName, userObjectID, clusterAdminRoleID)
+	if err != nil {
+		return fmt.Errorf("failed to assign cluster admin role: %w", err)
+	}
+
+	// Validate assignment
+	err = CheckClusterAdminPermissions(ctx, kubeconfigPath)
+	if err == nil {
+		return nil
+	}
+
+	// Wait for role assignment to be effective
+	fmt.Println("Wait for role assignment to be effective")
+	timeout := time.After(options.Timeout)
+	ticker := time.NewTicker(options.CheckFrequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-timeout:
+			return fmt.Errorf("timed out waiting for role assignment to be effective")
+		case <-ticker.C:
+			err = CheckClusterAdminPermissions(ctx, kubeconfigPath)
+			if err == nil {
+				fmt.Println("Cluster admin permissions are now effective")
+				return nil
+			}
+			fmt.Println("Waiting for role assignment to be effective...")
+		}
+	}
+}
+
+func CheckClusterAdminPermissions(ctx context.Context, kubeconfigPath string) error {
+	clientset, err := createKubeClient(kubeconfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to create Kubernetes client: %w", err)
+	}
+
+	// Implement the logic to test cluster admin permissions
```

Comment:
Depending on how the kubeconfig is created, the API server may not necessarily be running at that time - not sure if you will hit this using AKS, but in the past I've seen a lot of flakes around this kind of logic that did not handle network flakes or server being temporarily down.

Reply:
for our scenario the kubeapi was always accessible once our bicep templates completed